apps/opencs/view/render/terrainselection.cpp

```diff
@@ -0,0 +1,251 @@
+
+    // Primitive
+    osg::ref_ptr<osg::DrawArrays> drawArrays = new osg::DrawArrays{osg::PrimitiveSet::Mode::LINES};
+
+    geometry->addPrimitiveSet(drawArrays);
+    geometry->getOrCreateStateSet()->setMode(GL_LIGHTING, osg::StateAttribute::Values::OFF);
+
+    mGeode = new osg::Geode();
+    mGeode->addDrawable(geometry);
+
+    activate();
+}
+
+CSVRender::TerrainSelection::~TerrainSelection()
+{
+    deactivate();
+}
+
+std::vector<std::pair<int, int>> CSVRender::TerrainSelection::getTerrainSelection() const
+{
+    return mSelection;
+}
+
+void CSVRender::TerrainSelection::selectTerrainTexture(const WorldspaceHitResult& hit)
+{
+    mSelection.clear();
+
+    const std::pair<int, int> localPos {toTextureCoords(hit.worldPos)};
+
+    if (std::find(mSelection.begin(), mSelection.end(), localPos) == mSelection.end())
+        mSelection.push_back(localPos);
+
+    update();
+}
+
+void CSVRender::TerrainSelection::onlyAddSelect(const WorldspaceHitResult& hit)
+{
+    const std::pair<int, int> localPos {toTextureCoords(hit.worldPos)};
+
+    if (std::find(mSelection.begin(), mSelection.end(), localPos) == mSelection.end())
+            mSelection.push_back(localPos);
+
+    update();
+}
+
+void CSVRender::TerrainSelection::toggleSelect(const WorldspaceHitResult& hit)
+{
+    const std::pair<int, int> localPos {toTextureCoords(hit.worldPos)};
+    const auto iter = std::find(mSelection.begin(), mSelection.end(), localPos);
+
+    if (iter != mSelection.end()) {
+        mSelection.erase(iter);
+    }
+    else {
+        mSelection.push_back(localPos);
+    }
+    update();
+}
+
+void CSVRender::TerrainSelection::activate()
+{
+    mParentNode->addChild(mGeode);
+}
+
+void CSVRender::TerrainSelection::deactivate()
+{
+    mParentNode->removeChild(mGeode);
+}
+
+std::pair<int, int> CSVRender::TerrainSelection::toTextureCoords(osg::Vec3d worldPos) const
+{
+    const double xd {worldPos.x() * landTextureSize / cellSize+(landTextureSize / cellSize)/4};
+    const double yd {worldPos.y() * landTextureSize / cellSize+(landTextureSize / cellSize)/4};
+
+    const auto x = static_cast<int>(std::floor(xd));
+    const auto y = static_cast<int>(std::floor(yd));
+
+    return std::make_pair(x, y);
+}
+
+std::pair<int, int> CSVRender::TerrainSelection::toVertexCoords(osg::Vec3d worldPos) const
+{
+    const double xd {worldPos.x() * landTextureSize / cellSize+(landTextureSize / cellSize)/4};
+    const double yd {worldPos.y() * landSize / cellSize+(landTextureSize / cellSize)/4};
+
+    const auto x = static_cast<int>(std::floor(xd));
+    const auto y = static_cast<int>(std::floor(yd));
+
+    return std::make_pair(x, y);
+}
+
+double CSVRender::TerrainSelection::toWorldCoords(int pos)
+{
+    return cellSize * static_cast<double>(pos) / landTextureSize;
+}
+
+size_t CSVRender::TerrainSelection::landIndex(int x, int y)
+{
+    return y * landSize + x;
+}
+
+void CSVRender::TerrainSelection::update()
```

Comment:
This is a pretty long function with lots of repetition in it. Might be worth looking at again for a bit of refectoring/simplification.